tuta-sdk/rust/sdk/src/crypto/kyber.rs

```diff
@@ -97,25 +104,57 @@ impl KyberPublicKey {
 	pub fn encapsulate(&self) -> KyberEncapsulation {
 		use pqcrypto_traits::kem::*;
 
-		let (shared_secret, ciphertext) = kyber1024_encapsulate(&self.public_key);
+		let (unbound_shared_secret, ciphertext) = mlkem1024_encapsulate(&self.public_key);
+
+		// turn ML-KEM into Kyber (except for some randomness hashing that we do not care about)
+		let shared_secret = bind_shared_secret_to_ciphertext(unbound_shared_secret, ciphertext);
 
 		KyberEncapsulation {
 			ciphertext: KyberCiphertext(ciphertext.as_bytes().try_into().unwrap()),
-			shared_secret: KyberSharedSecret(shared_secret.as_bytes().try_into().unwrap()),
+			shared_secret,
 		}
 	}
 }
 
-impl From<PQCryptoKyber1024PublicKey> for KyberPublicKey {
-	fn from(value: PQCryptoKyber1024PublicKey) -> Self {
+/// This is a redundant step to bind the derived shared secret to the ciphertext.
+/// It was part of the original round 3 Kyber submission specification and the reference implementation.
+/// It was removed from the NIST ML-KEM draft for efficiency because the re-encryption step in decapsulation prevents any attacks.
+/// Therefore, pqcrypto updated the implementation, and we keep this step for compatibility in order to avoid rolling out a new protocol version.
+/// returns KDF(unbound_shared_secret || H(ciphertext))
+/// @return a shared_secret that is bound to the ciphertext and derived from the unbound_shared_secret
+fn bind_shared_secret_to_ciphertext(
+	unbound_shared_secret: PQCryptoMlKem1024SharedSecret,
+	ciphertext: PQCryptoMlKem1024Ciphertext,
+) -> KyberSharedSecret {
+	let hashed_ciphertext = sha::sha3_256(ciphertext.as_bytes());
+	let kdf_input = [
+		unbound_shared_secret.as_bytes(),
+		hashed_ciphertext.as_slice(),
+	]
+	.concat();
```

Comment:
we could avoid concating them if we provide both to the hasher one after another. Not sure if it's worth it though (we can make a macro :D)

Reply:
I pushed a commit to address this. what do you think?

Reply:
I think it's better :)